pkg/controllers/deprovisioning/expiration.go

```diff
@@ -0,0 +1,158 @@
+	return e.clock.Now().After(getExpirationTime(n.Node, provisioner))
+}
+
+// SortCandidates orders expired nodes by when they've expired
+func (e *Expiration) SortCandidates(nodes []CandidateNode) []CandidateNode {
+	sort.Slice(nodes, func(i int, j int) bool {
+		return getExpirationTime(nodes[i].Node, nodes[i].provisioner).Before(getExpirationTime(nodes[j].Node, nodes[j].provisioner))
+	})
+	return nodes
+}
+
+// ComputeCommand generates a deprovisioning command given deprovisionable nodes
+func (e *Expiration) ComputeCommand(ctx context.Context, attempt int, candidates ...CandidateNode) (Command, error) {
+	pdbs, err := NewPDBLimits(ctx, e.kubeClient)
+	if err != nil {
+		return Command{action: actionFailed}, fmt.Errorf("tracking PodDisruptionBudgets, %w", err)
+	}
+	// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
+	// scheduling unless its really needed
+	if !canBeTerminated(candidates[attempt], pdbs) {
+		return Command{action: actionNotPossible}, nil
+	}
+
+	// Only expire one node at a time.
+	node := candidates[attempt]
+	// Check if we need to create any nodes.
+	// Ignore if all pods cannot schedule. Node is expired and we're just calculating if we need to spin up a new node.
+	newNodes, _, err := simulateScheduling(ctx, e.kubeClient, e.cluster, e.provisioner, node)
+	if err != nil {
+		// if a candidate node is now deleting, just retry
+		if errors.Is(err, errCandidateNodeDeleting) {
+			return Command{action: actionDoNothing}, nil
+		}
+		return Command{}, err
+	}
+
+	// were we able to schedule all the pods on the inflight nodes?
+	if len(newNodes) == 0 {
+		return Command{
+			nodesToRemove: []*v1.Node{node.Node},
+			action:        actionDelete,
+			created:       e.clock.Now(),
+		}, nil
+	}
+
+	// get the current node price based on the offering
+	// fallback if we can't find the specific zonal pricing data
+	offering, ok := cloudprovider.GetOffering(node.instanceType, node.capacityType, node.zone)
+	if !ok {
+		return Command{action: actionFailed}, fmt.Errorf("getting offering price from candidate node, %w", err)
+	}
+	newNodes[0].InstanceTypeOptions = filterByPrice(newNodes[0].InstanceTypeOptions, newNodes[0].Requirements, offering.Price)
+	if len(newNodes[0].InstanceTypeOptions) == 0 {
+		// no instance types remain after filtering by price
+		return Command{action: actionNotPossible}, nil
+	}
+
+	return Command{
+		nodesToRemove:    []*v1.Node{node.Node},
+		action:           actionReplace,
+		replacementNodes: newNodes,
+		created:          e.clock.Now(),
+	}, nil
+}
+
+// ValidateCommand validates a command for a deprovisioner
+// We don't need to do another scheduling simulation since TTL is 0 seconds.
+func (e *Expiration) ValidateCommand(ctx context.Context, candidates []CandidateNode, cmd Command) (bool, error) {
+	// Only expire one node at a time
```

Comment:
Yeah you're right. I was planning on adding this and then pulling ValidateCommand out of the interface since after thinking it seems like only Consolidation needs a TTL

Reply:
Agree with @tzneal, can we simplify with a `return true, nil` and then create a TODO for the removal